### DISCUSSION TITLE
Multiple queries, custom return data?

### DISCUSSION THREAD

**Discussion by steelzeh** [pocketbase/pocketbase#1334] [Q&A] [answered]
I have 2 tables, `group` and `group_member`, `group_member` contains `group_id` and `user_id`. 
I'm fetching all items from `group_member` where `user_id` matches the authenticated user id.
Then I'm enriching the record with the `group`.

So the JSON being returned, returns the group_member and the group. All of this works fine.
But let's say i want to join in something that is not part of the record I'm fetching, like lets say i want to fetch all of the users that are in each group.

So if i fetch a group for the current authenticated user, let's call that `Group A`. My current code will return group_member and expand with `Group A`, but i would also like to find ALL the other users that are members of `Group A` and return them inside the JSON somewhere, probably in the expand.

Code looks like this:

```
app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.AddRoute(echo.Route{
			Method: http.MethodGet,
			Path:   "/api/custom/groups",
			Handler: func(c echo.Context) error {
				authRecord, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
				if authRecord == nil {
					return apis.NewForbiddenError("Only auth records can access this endpoint", nil)
				}

				records, err := app.Dao().FindRecordsByExpr("group_member",
					dbx.NewExp("user_id = {:id}", dbx.Params{"id": authRecord.Id}),
				)
				if err != nil {
					return apis.NewNotFoundError("", err)
				}

				apis.EnrichRecords(c, app.Dao(), records, "group_id")

				return c.JSON(http.StatusOK, records)
			},
			Middlewares: []echo.MiddlewareFunc{
				apis.ActivityLogger(app),
			},
		})

		return nil
	})
```
this returns this:

```
[
    {
        "collectionId": "v410f4jbpop4pu5",
        "collectionName": "group_member",
        "created": "2022-12-21 13:47:00.975Z",
        "expand": {
            "group_id": {
                .....data.....
            }
        },
        "group_id": "qrk2go1sqa9grcs",
        "id": "3tdwn0daduf380j",
        "is_author": true,
        "updated": "2022-12-21 13:48:30.394Z",
        "user_id": "v7sfstr3teqzn0e"
    }
]
```

I would like to also return an array of all users in the group. How do i go finding and joining all users for the group.
SQL would be something like this: `SELECT * FROM group_member WHERE group_id = "qrk2go1sqa9grcs";` so ideally i would have to run this select for each group record?

**Comment by ganigeorgiev** [comment, marked as answer]
_I'm not sure that I fully understand the requirements and the described data structure._

---

If the requirement is to return all groups where the authenticated user has `group_member` relation, then I personally wouldn't use a custom endpoint for that and instead send a request to the `group` list/search endpoint:

```
GET /api/collections/group/records?expand=group_member(group_id).user_id&perPage=50
```

The above will return a `group` records list (paginated) with **indirect** expanded `group_id` and `group_id.user_id` relations.
The access controls can be done via each collections API rules (only those records that satisfies their collection "View API rule" will be expanded). If you haven't already, you can check https://pocketbase.io/docs/expanding-relations/.

---

If you want to go with the custom query approach, then there are several ways you can get all users for the group_member's group.

One workaround would be to pass as default expand to the `EnrichRecords` - `group_id.group_member.user_id` (this will expand all groups, each with their group member, each with their user; the same "View API rule" expand requirement applies).

Another approach could be to send a new query for `group_members` and then assign them manually to each of the previous records. Something like:

```go
// extract the group ids
var groupIds = make([]any, len(records))
for i, r := range records {
	groupIds[i] = r.GetString("group_id")
}

// find all group members by their group_id
groupMembers, err := app.Dao().FindRecordsByExpr("group_member", dbx.In("group_id", groupIds...))
if err != nil {
	return apis.NewBadRequestError("Failed to fetch other group members", err)
}

// index the related group_member by their group_id
indexedGroupMembers := map[string][]*models.Record{}
for _, m := range groupMembers {
	indexedGroupMembers[m.GetString("group_id")] = append(indexedGroupMembers[m.GetString("group_id")], m)
}

// add the groupped members to each of the initial records
for _, r := range records {
	r.WithUnkownData(true) // allow serializing custom/unknown fields
	r.Set("other_group_members", indexedGroupMembers[r.GetString("group_id")])
}

return c.JSON(http.StatusOK, records)
```

Third approach would be to construct an entirely custom query using `app.Dao().DB().NewQuery(...).All()` but this is more involving and will require to define upfront a special struct so that you can map the db fields.

**Reply by steelzeh** [reply]
Thank you very much @ganigeorgiev, the code does exactly what i need it to do.

Had a hard time figuring out how to do it my self, great with examples. Especially the `WithUnkownData(true)`, was unsure of how to add custom fields.
